Update unit-testing-with-dotnet-test.md (#15954)
dotnet/docs: docs/core/testing/unit-testing-with-dotnet-test.md

---
# Unit testing C# in .NET Core using dotnet test and xUnit

This tutorial shows how to build a solution containing a unit test project and source code project. To follow the tutorial using a pre-built solution, [view or download the sample code](https://github.com/dotnet/samples/tree/master/core/getting-started/unit-testing-using-dotnet-test/). For download instructions, see [Samples and Tutorials](../../samples-and-tutorials/index.md#viewing-and-downloading-samples).

## Create the solution

In this section, a solution is created that contains the source and test projects. The completed solution has the following directory structure:

```
/unit-testing-using-dotnet-test
    unit-testing-using-dotnet-test.sln
    /PrimeService
        PrimeService.cs
        PrimeService.csproj
    /PrimeService.Tests
        PrimeService_IsPrimeShould.cs
        PrimeServiceTests.csproj

> ```

Reviewer: This could be confusing to beginners. The full folder structure (tree) should come after the instructions that creates the solution and the projects with all this files.

Author: I disagree. As they go on they can refer to that.

The following instructions provide the steps to create the test solution. See [Commands to create test solution](#create-test-cmd) for instructions to create the test solution in one step.

* Open a shell window.
* Run the following command:

  ```dotnetcli
  dotnet new sln -o unit-testing-using-dotnet-test
  ```

  The [`dotnet new sln`](../tools/dotnet-new.md) command creates a new solution in the *unit-testing-using-dotnet-test* directory.

> * Change directories to the *unit-testing-using-dotnet-test* folder.

Reviewer: ```suggestion
* Change directory to the *unit-testing-using-dotnet-test* folder.
```

  
  ```csharp
    using System;

    namespace Prime.Services
    {
        public class PrimeService
        {
            public bool IsPrime(int candidate)
            {
                throw new NotImplementedException("Not implemented.");
            }
        }
    }
  ```

* The preceding code:
  * Throws a <xref:System.NotImplementedException> with a message indicating it's not implemented.
  * Is updated later in the tutorial.

<!-- preceding code shows an english bias. Message makes no sense outside english -->

* In the *unit-testing-using-dotnet-test* directory, run the following command to add the class library project to the solution:

  ```dotnetcli
  dotnet sln add ./PrimeService/PrimeService.csproj
  ```

* Create the *PrimeService.Tests* project by running the following command:

  ```dotnetcli
  dotnet new xunit -o PrimeService.Tests
  ```

* The preceding command:
  * Creates the *PrimeService.Tests* project in the *PrimeService.Tests* directory. The test project uses [xUnit](https://xunit.github.io/) as the test library.
  * Configures the test runner by adding the following `<PackageReference />`elements to the project file:

>      * "Microsoft.NET.Test.Sdk"

Reviewer: ```suggestion
    * "Microsoft.NET.Test.Sdk"
```

Author: No, what I have is correct.

>      * "xunit"

Reviewer: ```suggestion
    * "xunit"
```

Author: No, this is correct. It needs to be the next level indented.

Reviewer: @Rick-Anderson, I removed only one space. so it's indented 4 spaces instead of 5. The linter was showing an error because of the 5 spaces.

>      * "xunit.runner.visualstudio"

Reviewer: ```suggestion
    * "xunit.runner.visualstudio"
```

Author: @Youssef1313 thanks, made it 4

* Add the test project to the solution file by running the following command:

  ```dotnetcli
  dotnet sln add ./PrimeService.Tests/PrimeService.Tests.csproj
  ```

* Add the `PrimeService` class library as a dependency to the *PrimeService.Tests* project:

  ```dotnetcli
  dotnet add ./PrimeService.Tests/PrimeService.Tests.csproj reference ./PrimeService/PrimeService.csproj  
  ```

> <a name="create-test-cmd"></a>

Reviewer: I wouldn't use anchors for linking. The docs.microsoft.com already gives the headers an id.

Author: But that heading is likely to change. I do this frequently in https://github.com/aspnet/AspNetCore.Docs

> ### Commands to create test solution

Reviewer: For me, "test" solution is confusing a little bit. The solution holds both projects, the test and the class lib.

Author: LOL, now you see why I added an anchor. Only one thing to change.

Author: Fixed. good suggestion.

> This section contains all the commands in the previous section. Skip this section if you've created the test solution.

Reviewer 2: I think this is clearer.

```suggestion
This section summarizes all the commands in the previous section. Skip this section if you've completed the steps in the previous section.
```

Author: Much better, thanks for the suggestion.

> The following commands create the test solution on a windows machine. For MacOS and unix, update the `ren` command to the OS version of `ren` to rename a file:

Reviewer: ```suggestion
The following commands create the test solution on a windows machine. For macOS and Unix, update the `ren` command to the OS version of `ren` to rename a file:
```

cd unit-testing-using-dotnet-test
dotnet new classlib -o PrimeService
ren .\PrimeService\Class1.cs PrimeService.cs
dotnet sln add ./PrimeService/PrimeService.csproj
dotnet new xunit -o PrimeService.Tests
dotnet add ./PrimeService.Tests/PrimeService.Tests.csproj reference ./PrimeService/PrimeService.csproj
dotnet sln add ./PrimeService.Tests/PrimeService.Tests.csproj
```

Follow the instructions for "Replace the code in *PrimeService.cs* with the following code" in the previous section.

## Create a test

A popular approach in test driven development (TDD) is to write a test before implementing the target code. This tutorial uses the TDD approach. The `IsPrime` method is callable, but not implemented. A test call to `IsPrime` fails. With TDD, a test is written that is known to fail. The target code is updated to make the test pass. You keep repeating this approach, writing a failing test and then updating the target code to pass.

Update the *PrimeService.Tests* project:

* Delete *PrimeService.Tests/UnitTest1.cs*.
* Create a *PrimeService.Tests/PrimeService_IsPrimeShould.cs*  file.
* Replace the code in *PrimeService_IsPrimeShould.cs* with the following code:

> ```csharp

Reviewer: This code should be indented to be under the list item.

Author: The doc build looks much better if it's not, ditto for the information after the block.

- `[Theory]` represents a suite of tests that execute the same code but have different input arguments.

- `[InlineData]` attribute specifies values for those inputs.

Rather than creating new tests, apply the preceding xUnit attributes to create a single theory. Replace the following code:

```csharp
[Fact]
public void IsPrime_InputIs1_ReturnFalse()
{
    var result = _primeService.IsPrime(1);

    Assert.False(result, "1 should not be prime");
}
```

with the following code:

[!code-csharp[Sample_TestCode](../../../samples/core/getting-started/unit-testing-using-dotnet-test/PrimeService.Tests/PrimeService_IsPrimeShould.cs?name=Sample_TestCode)]

In the preceding code, `[Theory]` and `[InlineData]` enable testing several values less than two. Two is the smallest prime number.

Run `dotnet test`, two of the tests fail. To make all of the tests pass, update the `IsPrime` method with the following code:

```csharp
public bool IsPrime(int candidate)
{
    if (candidate < 2)
    {
        return false;
    }
    throw new NotImplementedException("Not fully implemented.");
}
```

> Following the TDD approach, add more failing test, then update the target code. See the [finished version of the tests](https://github.com/dotnet/samples/blob/master/core/getting-started/unit-testing-using-dotnet-test/PrimeService.Tests/PrimeService_IsPrimeShould.cs) and the [complete implementation of the library](https://github.com/dotnet/samples/blob/master/core/getting-started/unit-testing-using-dotnet-test/PrimeService/PrimeService.cs).

Reviewer 3: "add more failing tests"
(missing 's')

Reviewer: @ardalis, opened https://github.com/dotnet/docs/pull/16093
